data/src/main/java/org/apache/iceberg/data/TableMigrationUtil.java

```diff
@@ -140,28 +141,36 @@ public static List<DataFile> listPartition(
             index -> {
               Metrics metrics = getAvroMetrics(fileStatus.get(index).getPath(), conf);
               datafiles[index] =
-                  buildDataFile(fileStatus.get(index), partitionValues, spec, metrics, "avro");
+                  metrics.recordCount() > 0
+                      ? buildDataFile(fileStatus.get(index), partitionValues, spec, metrics, "avro")
+                      : null;
             });
       } else if (format.contains("parquet")) {
         task.run(
             index -> {
               Metrics metrics =
                   getParquetMetrics(fileStatus.get(index).getPath(), conf, metricsSpec, mapping);
               datafiles[index] =
-                  buildDataFile(fileStatus.get(index), partitionValues, spec, metrics, "parquet");
+                  metrics.recordCount() > 0
+                      ? buildDataFile(
+                          fileStatus.get(index), partitionValues, spec, metrics, "parquet")
+                      : null;
             });
       } else if (format.contains("orc")) {
         task.run(
             index -> {
               Metrics metrics =
                   getOrcMetrics(fileStatus.get(index).getPath(), conf, metricsSpec, mapping);
               datafiles[index] =
-                  buildDataFile(fileStatus.get(index), partitionValues, spec, metrics, "orc");
+                  metrics.recordCount() > 0
```

Comment:
I checked that the "number of rows" stat is required for Parquet files, however, for me it seems that it is optional for ORC (haven't checked Avro as I don't have much knowledge on that area.) https://github.com/apache/orc/blob/main/proto/orc_proto.proto#L360 So if my assumption is correct then we skip files if the reader for some reason didn't write the number of rows metric into the footer even though there are stripes/rows in the file.